airflow/decorators/assets.py

```diff
@@ -0,0 +1,155 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import inspect
+import types
+from functools import wraps
+from typing import TYPE_CHECKING, Any, Callable, Iterator, Mapping, cast
+
+import attrs
+
+from airflow.assets import Asset, AssetRef
+from airflow.models.asset import _fetch_active_assets_by_name
+from airflow.models.dag import DAG, ScheduleArg
+from airflow.providers.standard.operators.python import PythonOperator
+
+if TYPE_CHECKING:
+    from typing import Sequence
+
+    from airflow.io.path import ObjectStoragePath
+
+
+class _AssetMainOperator(PythonOperator):
+    def __init__(self, *, definition_name: str, uri: str | None = None, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self._definition_name = definition_name
+        self._uri = uri
+
+    def _iter_kwargs(
+        self, context: Mapping[str, Any], active_assets: dict[str, Asset]
+    ) -> Iterator[tuple[str, Any]]:
+        value: Any
+        for key in inspect.signature(self.python_callable).parameters:
+            if key == "self":
+                key = "_self"
+                value = active_assets.get(self._definition_name)
+                if not value:
+                    value = Asset(name=self._definition_name)
+                    if self._uri is not None:
+                        value.uri = self._uri
+            elif key == "context":
+                value = context
+            else:
+                value = active_assets.get(key, Asset(name=key))
+            yield key, value
+
+    def determine_kwargs(self, context: Mapping[str, Any]) -> Mapping[str, Any]:
+        active_assets: dict[str, Asset] = {}
+        asset_names = [asset_ref.name for asset_ref in self.inlets if isinstance(asset_ref, AssetRef)]
+        if "self" in inspect.signature(self.python_callable).parameters:
+            asset_names.append(self._definition_name)
+
+        if asset_names:
+            active_assets = _fetch_active_assets_by_name(asset_names)
+        return dict(self._iter_kwargs(context, active_assets))
+
+
+def _handle_self_argument(func: types.FunctionType) -> types.FunctionType:
+    @wraps(func)
+    def wrapper(_self: Any, *args: Sequence[Any], **kwargs: dict[str, Any]) -> Any:
+        return func(_self, *args, **kwargs)
+
+    return cast(types.FunctionType, wrapper)
+
+
+@attrs.define(kw_only=True)
+class AssetDefinition(Asset):
+    """
+    Asset representation from decorating a function with ``@asset``.
+
+    :meta private:
+    """
+
+    function: types.FunctionType
+    schedule: ScheduleArg
+
+    def __attrs_post_init__(self) -> None:
+        parameters = inspect.signature(self.function).parameters
+        if "self" in parameters:
+            self.function: types.FunctionType = _handle_self_argument(cast(types.FunctionType, self.function))  # type: ignore[assignment]
```

Comment:
Why do we need to do `cast(..., self.function)`? It is declared as `types.FunctionType`…

Reply:
And why does `function` need to be `FunctionType` at all, not just `Callable`?

Reply:
> Why do we need to do cast(..., self.function)? It is declared as types.FunctionType…

It's be classified as ClassDef by mypy. I "guess" it's due to the `self`

> And why does function need to be FunctionType at all, not just Callable?

It used to be `FunctionType`. I thought that was by design. 🤔 I can change it to `Callable`.

Reply:
It’s sort of by design because this should only be able to decorate a real function, not any callable (say lambda). But it’s not really that big a deal if the typing causes too many troubles elsewhere.

Reply:
Sounds good. I already make it `Callable` and this part seems to work well